Content.Shared/Mining/MiningScannerSystem.cs

```diff
@@ -0,0 +1,125 @@
+    {
+        if (_container.TryGetContainingContainer((ent.Owner, null, null), out var container))
+            UpdateViewerComponent(container.Owner);
+    }
+
+    public void UpdateViewerComponent(EntityUid uid)
+    {
+        // Don't do updates on non-mobs.
+        if (!HasComp<MobStateComponent>(uid))
+            return;
+
+        Entity<MiningScannerComponent>? scannerEnt = null;
+
+        if (_inventory.TryGetContainerSlotEnumerator(uid, out var enumerator))
+        {
+            while (enumerator.MoveNext(out var slot))
+            {
+                foreach (var ent in slot.ContainedEntities)
+                {
+                    if (!TryComp<MiningScannerComponent>(ent, out var scannerComponent) ||
+                        !TryComp<ItemToggleComponent>(ent, out var toggle))
+                        continue;
+
+                    if (!toggle.Activated)
+                        continue;
+
+                    if (scannerEnt == null || scannerComponent.Range > scannerEnt.Value.Comp.Range)
+                       scannerEnt = (ent, scannerComponent);
+                }
+            }
+        }
+
+        foreach (var hand in _hands.EnumerateHands(uid))
+        {
+            if (!TryComp<MiningScannerComponent>(hand.HeldEntity, out var scannerComponent) ||
+                !TryComp<ItemToggleComponent>(hand.HeldEntity, out var toggle))
+                continue;
+
+            if (!toggle.Activated)
+                continue;
+
+            if (scannerEnt == null || scannerComponent.Range > scannerEnt.Value.Comp.Range)
+                scannerEnt = (hand.HeldEntity.Value, scannerComponent);
+        }
+
+        if (_net.IsServer)
+        {
+            if (scannerEnt == null)
+            {
+                if (TryComp<MiningScannerViewerComponent>(uid, out var viewer))
+                    viewer.QueueRemoval = true;
+            }
+            else
+            {
+                var viewer = EnsureComp<MiningScannerViewerComponent>(uid);
+                viewer.ViewRange = scannerEnt.Value.Comp.Range;
+                viewer.QueueRemoval = false;
+                viewer.NextPingTime = _timing.CurTime + viewer.PingDelay;
+                Dirty(uid, viewer);
+            }
+        }
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<MiningScannerViewerComponent, TransformComponent>();
+        while (query.MoveNext(out var uid, out var viewer, out var xform))
+        {
+            if (viewer.QueueRemoval)
+            {
+                RemCompDeferred(uid, viewer);
+                continue;
+            }
+
+            if (_timing.CurTime < viewer.NextPingTime)
+                continue;
+
+            viewer.NextPingTime = _timing.CurTime + viewer.PingDelay;
+            viewer.LastPingLocation = xform.Coordinates;
+            _audio.PlayEntity(viewer.PingSound, uid, uid);
```

Comment:
Should this be playpredicted?

Reply:
no, because it should only be playing for the person holding the scanner. i added some extra checks, though.

Reply:
You can PlayLocal then if client's predicting it.

Reply:
play local doesn't seem to exist

Reply:
mfw no popup equivalent someone should code that